src/Map.ts

```diff
@@ -0,0 +1,825 @@
+        )
+      }
+      return r
+    },
+    empty
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+function filter<K, A, B extends A>(fa: Map<K, A>, p: Refinement<A, B>): Map<K, B>
+function filter<K, A>(fa: Map<K, A>, p: Predicate<A>): Map<K, A>
+function filter<K, A>(fa: Map<K, A>, p: Predicate<A>): Map<K, A> {
+  return filterWithKey(fa, (_, a) => p(a))
+}
+
+/**
+ * @since 1.14.0
+ */
+const mapWithKey = <K, A, B>(fa: Map<K, A>, f: (k: K, a: A) => B): Map<K, B> => {
+  const m = new Map<K, B>()
+  const entries = fa.entries()
+  let e: IteratorResult<[K, A]>
+  while (!(e = entries.next()).done) {
+    const [key, a] = e.value
+    m.set(key, f(key, a))
+  }
+  return m
+}
+
+/**
+ * @since 1.14.0
+ */
+const map_ = <K, A, B>(fa: Map<K, A>, f: (a: A) => B): Map<K, B> => mapWithKey(fa, (_, a) => f(a))
+
+/**
+ * @since 1.14.0
+ */
+const reduce = <K>(O: Ord<K>): (<A, B>(fa: Map<K, A>, b: B, f: (b: B, a: A) => B) => B) => {
+  return (fa, b, f) => reduceWithKey(O)(fa, b, (_, b, a) => f(b, a))
+}
+
+/**
+ * @since 1.14.0
+ */
+const foldMap = <K>(O: Ord<K>): (<M>(M: Monoid<M>) => <A>(fa: Map<K, A>, f: (a: A) => M) => M) => M => (fa, f) =>
+  foldMapWithKey(O)(M)(fa, (_, a) => f(a))
+
+/**
+ * @since 1.14.0
+ */
+const foldr = <K>(O: Ord<K>): (<A, B>(fa: Map<K, A>, b: B, f: (a: A, b: B) => B) => B) => {
+  return (fa, b, f) => foldrWithKey(O)(fa, b, (_, a, b) => f(a, b))
+}
+
+/**
+ * @since 1.14.0
+ */
+const reduceWithKey = <K>(O: Ord<K>): (<A, B>(fa: Map<K, A>, b: B, f: (k: K, b: B, a: A) => B) => B) => {
+  const keysO = keys(O)
+  return <A, B>(fa: Map<K, A>, b: B, f: (k: K, b: B, a: A) => B): B => {
+    let out: B = b
+    const ks = keysO(fa)
+    const len = ks.length
+    for (let i = 0; i < len; i++) {
+      const k = ks[i]
+      out = f(k, out, fa.get(k)!)
+    }
+    return out
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const foldMapWithKey = <K>(O: Ord<K>): (<M>(M: Monoid<M>) => <A>(fa: Map<K, A>, f: (k: K, a: A) => M) => M) => {
+  const keysO = keys(O)
+  return <M>(M: Monoid<M>) => <A>(fa: Map<K, A>, f: (k: K, a: A) => M): M => {
+    let out: M = M.empty
+    const ks = keysO(fa)
+    const len = ks.length
+    for (let i = 0; i < len; i++) {
+      const k = ks[i]
+      out = M.concat(out, f(k, fa.get(k)!))
+    }
+    return out
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+const foldrWithKey = <K>(O: Ord<K>): (<A, B>(fa: Map<K, A>, b: B, f: (k: K, a: A, b: B) => B) => B) => {
+  const keysO = keys(O)
+  return <A, B>(fa: Map<K, A>, b: B, f: (k: K, a: A, b: B) => B): B => {
+    let out: B = b
+    const ks = keysO(fa)
+    const len = ks.length
+    for (let i = len - 1; i >= 0; i--) {
+      const k = ks[i]
+      out = f(k, fa.get(k)!, out)
+    }
+    return out
+  }
+}
+
+/**
+ * Create a Map with one key/value pair
+ *
+ * @since 1.14.0
+ */
+export const singleton = <K, A>(k: K, a: A): Map<K, A> => {
+  return new Map<K, A>([[k, a]])
+}
+
+/**
+ * @since 1.14.0
+ */
+function traverseWithKey<F extends URIS3>(
```

Comment:
overloadings can be removed

Reply:
Same as before, can you explain why? Are they only needed when treated as standalone functions & don't exist within a type class instance?

Reply:
I made this change. Marking as resolved.